precompiles/outposts/osmosis/tx.go

```diff
@@ -32,12 +35,112 @@ const (
 // Swap is a transaction that swap tokens on the Osmosis chain using
 // an ICS20 transfer with a custom memo field to trigger the XCS V2 contract.
 func (p Precompile) Swap(
-	_ sdk.Context,
-	_ common.Address,
-	_ vm.StateDB,
-	_ *vm.Contract,
-	_ *abi.Method,
-	_ []interface{},
+	ctx sdk.Context,
+	origin common.Address,
+	stateDB vm.StateDB,
+	contract *vm.Contract,
+	method *abi.Method,
+	args []interface{},
 ) ([]byte, error) {
-	return nil, nil
+	sender, input, output, amount, slippagePercentage, windowSeconds, receiver, err := ParseSwapPacketData(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// The provided sender address should always be equal to the origin address.
+	// In case the contract caller address is the same as the sender address provided,
+	// update the sender address to be equal to the origin address.
+	// Otherwise, if the provided sender address is different from the origin address,
+	// return an error because is a forbidden operation
+	sender, err = ics20.CheckOriginAndSender(contract, origin, sender)
+	if err != nil {
+		return nil, err
+	}
+
+	inputDenom, err := p.erc20Keeper.GetTokenDenom(ctx, input)
+	if err != nil {
+		return nil, err
+	}
+	outputDenom, err := p.erc20Keeper.GetTokenDenom(ctx, output)
+	if err != nil {
+		return nil, err
+	}
+
+	// We need the bonded denom just for the outpost alpha version where the
+	// the only two inputs allowed are aevmos and uosmo.
+	bondDenom := p.stakingKeeper.GetParams(ctx).BondDenom
+
+	err = ValidateInputOutput(inputDenom, outputDenom, bondDenom, p.portID, p.channelID)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the receiver has not the prefix "osmo", we should compute its address
+	// in the Osmosis chain as a recovery address for the contract.
+	onFailedDelivery := CreateOnFailedDeliveryField(receiver)
```

Comment:
Not sure if this is needed, the `receiver` bech32 will always need to be present and a valid `osmo` bech32 otherwise it needs to error earlier in the call and we can populate `on_failed_delivery` with that bech32 always.

Reply:
Can't be any valid bech32 address the receiver?

Reply:
The `receiver` in the `memo` yes for any chain but the `receiver` you are passing is the same one you are giving to ICS20.